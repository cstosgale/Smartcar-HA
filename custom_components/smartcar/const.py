from enum import StrEnum, auto

DOMAIN = "smartcar"
DEFAULT_NAME = "Smartcar"
API_HOST = "https://api.smartcar.com"

PLATFORMS = [
    "sensor",
    "switch",
    "lock",
    "device_tracker",
    "binary_sensor",
    "number",
]

OAUTH2_AUTHORIZE = "https://connect.smartcar.com/oauth/authorize"
OAUTH2_TOKEN = "https://auth.smartcar.com/oauth/token"
SMARTCAR_MODE = "live"


class Scope(StrEnum):
    READ_VEHICLE_INFO = auto()
    READ_VIN = auto()
    READ_BATTERY = auto()
    READ_CHARGE = auto()
    READ_ENGINE_OIL = auto()
    READ_FUEL = auto()
    READ_LOCATION = auto()
    READ_ODOMETER = auto()
    READ_SECURITY = auto()
    READ_TIRES = auto()
    CONTROL_CHARGE = auto()
    CONTROL_SECURITY = auto()


REQUIRED_SCOPES = [
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
]

CONFIGURABLE_SCOPES = [scope for scope in Scope if scope not in REQUIRED_SCOPES]

DEFAULT_SCOPES = [
    Scope.READ_BATTERY,
    Scope.READ_CHARGE,
    Scope.READ_LOCATION,
    Scope.READ_ODOMETER,
    Scope.READ_SECURITY,
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
    Scope.CONTROL_CHARGE,
]
