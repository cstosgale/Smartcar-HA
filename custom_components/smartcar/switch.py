from dataclasses import dataclass
import logging

from homeassistant.components.switch import SwitchEntity, SwitchEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .coordinator import SmartcarVehicleCoordinator
from .entity import SmartcarEntity, SmartcarEntityDescription

_LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True, kw_only=True)
class SmartcarSwitchDescription(SwitchEntityDescription, SmartcarEntityDescription):
    """Class describing Smartcar switch entities."""


ENTITY_DESCRIPTIONS: tuple[SwitchEntityDescription, ...] = (
    SmartcarSwitchDescription(
        key="charging",
        name="Charging",
        value_key_path="charge.state",
        value_cast=lambda value: value == "CHARGING",
        icon="mdi:ev-plug-type2",
    ),
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up switches from coordinator."""
    coordinators: dict[str, SmartcarVehicleCoordinator] = (
        entry.runtime_data.coordinators
    )
    entities = []
    for vin, coordinator in coordinators.items():
        for description in ENTITY_DESCRIPTIONS:
            if coordinator.is_scope_enabled(description.key, verbose=True):
                entities.append(SmartcarChargingSwitch(coordinator, description))
    _LOGGER.info(f"Adding {len(entities)} Smartcar switch entities")
    async_add_entities(entities)


class SmartcarChargingSwitch(SmartcarEntity, SwitchEntity):
    _attr_has_entity_name = True

    @property
    def is_on(self):
        return self._extract_value()

    async def async_turn_on(self, **kwargs):
        if await self._async_send_command("/charge", {"action": "START"}):
            self._inject_raw_value("CHARGING")
            self.async_write_ha_state()

    async def async_turn_off(self, **kwargs):
        if await self._async_send_command("/charge", {"action": "STOP"}):
            self._inject_raw_value("NOT_CHARGING")
            self.async_write_ha_state()
