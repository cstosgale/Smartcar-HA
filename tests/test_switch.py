"""Test switch entities."""

from unittest.mock import AsyncMock

from aiohttp import ClientResponseError
from homeassistant.components.switch import DOMAIN as SWITCH_DOMAIN
from homeassistant.const import (
    ATTR_ENTITY_ID,
    SERVICE_TURN_OFF,
    SERVICE_TURN_ON,
    STATE_OFF,
    STATE_ON,
)
from homeassistant.core import HomeAssistant
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from pytest_homeassistant_custom_component.test_util.aiohttp import AiohttpClientMocker
from syrupy.assertion import SnapshotAssertion

from . import MOCK_API_ENDPOINT, setup_integration

NO_ERROR = None.__class__


@pytest.mark.parametrize(
    (
        "service_action",
        "api_status",
        "api_status_slug",
        "expected_state",
        "expected_raises",
    ),
    [
        (SERVICE_TURN_ON, 200, "success", STATE_ON, NO_ERROR),
        (SERVICE_TURN_OFF, 200, "success", STATE_OFF, NO_ERROR),
        (SERVICE_TURN_ON, 409, "unreachable", STATE_OFF, NO_ERROR),
        (SERVICE_TURN_OFF, 409, "unreachable", STATE_OFF, NO_ERROR),
        (SERVICE_TURN_ON, 401, "unauthroized", STATE_OFF, NO_ERROR),
        (SERVICE_TURN_OFF, 401, "unauthroized", STATE_OFF, NO_ERROR),
        (SERVICE_TURN_OFF, 500, "server", STATE_OFF, ClientResponseError),
    ],
)
@pytest.mark.parametrize("vehicle_fixture", ["unknown_make"])
async def test_switch(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    service_action: str,
    api_status: int,
    api_status_slug: str,
    expected_state: str,
    expected_raises: Exception,
) -> None:
    """Test switching charging on/off."""

    await setup_integration(hass, mock_config_entry)
    assert len(aioclient_mock.mock_calls) == 1

    aioclient_mock.post(
        f"{MOCK_API_ENDPOINT}/v2.0/vehicles/{vehicle['id']}/charge",
        status=api_status,
        json={
            "message": "Some message related to the action unused by our code",
            "status": api_status_slug,
        },
    )

    try:
        await hass.services.async_call(
            SWITCH_DOMAIN,
            service_action,
            {ATTR_ENTITY_ID: "switch.smartcar_784n_charging"},
            blocking=True,
        )
    except Exception as error:  # noqa: BLE001
        raised_error = error
    else:
        raised_error = None  # type: ignore[assignment]

    switch_state = hass.states.get("switch.smartcar_784n_charging")
    assert switch_state.state == expected_state
    assert isinstance(
        raised_error,
        expected_raises,  # type: ignore[arg-type]
    )

    assert len(aioclient_mock.mock_calls) == 2
    assert [tuple(mock_call) for mock_call in aioclient_mock.mock_calls[1:]] == snapshot
