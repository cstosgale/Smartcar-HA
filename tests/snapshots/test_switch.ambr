# serializer version: 1
# name: test_switch[unknown_make-turn_off-200-success-off-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'STOP',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_off-401-unauthroized-off-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'STOP',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_off-409-unreachable-off-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'STOP',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_off-500-server-off-ClientResponseError]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'STOP',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_on-200-success-on-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'START',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_on-401-unauthroized-off-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'START',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_switch[unknown_make-turn_on-409-unreachable-off-NoneType]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge'),
      dict({
        'action': 'START',
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
